driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -337,28 +337,43 @@ private PoolingOptions options() {
       }
     }
 
+    Connection leastBusy = null;
+
     if (connections[shardId].isEmpty()) {
       if (host.convictionPolicy.canReconnectNow()) {
         if (connectionsPerShard == 0) {
           maybeSpawnNewConnection(shardId);
+          return enqueue(timeout, unit, maxQueueSize, shardId);
         } else if (scheduledForCreation[shardId].compareAndSet(0, connectionsPerShard)) {
           for (int i = 0; i < connectionsPerShard; i++) {
             // We don't respect MAX_SIMULTANEOUS_CREATION here because it's  only to
             // protect against creating connection in excess of core too quickly
             manager.blockingExecutor().submit(new ConnectionTask(shardId));
           }
+          return enqueue(timeout, unit, maxQueueSize, shardId);
         }
-        return enqueue(timeout, unit, maxQueueSize, shardId);
       }
-    }
-
-    int minInFlight = Integer.MAX_VALUE;
-    Connection leastBusy = null;
-    for (Connection connection : connections[shardId]) {
-      int inFlight = connection.inFlight.get();
-      if (inFlight < minInFlight) {
-        minInFlight = inFlight;
-        leastBusy = connection;
+      // connections for this shard are still being initialized so pick connection for any shard
```

Comment:
You walk _all_ shards to find a connection with least in-flight connections? Why? Wouldn't picking a shard by random (and searching for least busy connection) be sufficient to distribute connection borrowing across different shards?

Reply:
Indeed the Least Busy method creates problem in the original driver, because we may end up always picking the same one. This works well with a single client, but with multiple clients it can create adversarial conditions.

Random is better.

Reply:
Done.